coordinator/src/main/scala/filodb.coordinator/queryplanner/SingleClusterPlanner.scala

```diff
@@ -269,45 +305,149 @@ class SingleClusterPlanner(val dataset: Dataset,
+      case lp: ScalarVectorBinaryOperation => materializeScalarVectorBinOp(qContext, lp, forceInProcess)
+      case lp: LabelValues                 => materializeLabelValues(qContext, lp, forceInProcess)
+      case lp: LabelNames                  => materializeLabelNames(qContext, lp, forceInProcess)
+      case lp: TsCardinalities             => materializeTsCardinalities(qContext, lp, forceInProcess)
+      case lp: SeriesKeysByFilters         => materializeSeriesKeysByFilters(qContext, lp, forceInProcess)
+      case lp: ApplyMiscellaneousFunction  => materializeApplyMiscellaneousFunction(qContext, lp, forceInProcess)
+      case lp: ApplySortFunction           => materializeApplySortFunction(qContext, lp, forceInProcess)
+      case lp: ScalarVaryingDoublePlan     => materializeScalarPlan(qContext, lp, forceInProcess)
       case lp: ScalarTimeBasedPlan         => materializeScalarTimeBased(qContext, lp)
-      case lp: VectorPlan                  => materializeVectorPlan(qContext, lp)
+      case lp: VectorPlan                  => materializeVectorPlan(qContext, lp, forceInProcess)
       case lp: ScalarFixedDoublePlan       => materializeFixedScalar(qContext, lp)
-      case lp: ApplyAbsentFunction         => materializeAbsentFunction(qContext, lp)
-      case lp: ApplyLimitFunction          => materializeLimitFunction(qContext, lp)
-      case lp: ScalarBinaryOperation       => materializeScalarBinaryOperation(qContext, lp)
-      case lp: SubqueryWithWindowing       => materializeSubqueryWithWindowing(qContext, lp)
-      case lp: TopLevelSubquery            => materializeTopLevelSubquery(qContext, lp)
-      case lp: LabelCardinality            => materializeLabelCardinality(qContext, lp)
+      case lp: ApplyAbsentFunction         => materializeAbsentFunction(qContext, lp, forceInProcess)
+      case lp: ApplyLimitFunction          => materializeLimitFunction(qContext, lp, forceInProcess)
+      case lp: ScalarBinaryOperation       => materializeScalarBinaryOperation(qContext, lp, forceInProcess)
+      case lp: SubqueryWithWindowing       => materializeSubqueryWithWindowing(qContext, lp, forceInProcess)
+      case lp: TopLevelSubquery            => materializeTopLevelSubquery(qContext, lp, forceInProcess)
+      case lp: LabelCardinality            => materializeLabelCardinality(qContext, lp, forceInProcess)
       //case _                               => throw new BadQueryException("Invalid logical plan")
     }
   }
   // scalastyle:on cyclomatic.complexity
 
-  override def materializeBinaryJoin(qContext: QueryContext,
-                                     lp: BinaryJoin): PlanResult = {
-    val lhs = walkLogicalPlanTree(lp.lhs, qContext)
+  /**
+   * Returns the shards spanned by a LogicalPlan's data.
+   * @return an occupied Option iff shards could be determined for all leaf-level plans.
+   */
+  private def getShardSpanFromLp(lp: LogicalPlan,
+                                 qContext: QueryContext): Option[Set[Int]] = {
```

Comment:
The docs can convey that it will return a `Some` only iff the entire exec plan subtree for this logical plan can be pushed down to individual shards and concatenated, as opposed to an exec plan with distributed execution followed by some form of reduction at the top.

Reply:
I think you've described this separate function (which uses `getShardSpanFromLp`) (new comments have been added since you last looked):
```
  /**
   * Returns the set of shards to which this BinaryJoin can be pushed down.
   * When this function returns a Some, the argument BinaryJoin can be materialized for each shard
   *   in the result, and the concatenation of these shard-local joins will, on execute(), yield the
   *   same RangeVectors as the non-pushed-down plan.
   * See materializeBinaryJoinWithPushdown for details about this pushdown optimization.
   * @return an occupied Option iff it is valid to perform the a pushdown optimization on this BinaryJoin.
   */
  private def getBinaryJoinPushdownShards(qContext: QueryContext,
                                          bj: BinaryJoin): Option[Set[Int]]
```

`getShardSpanFromLp` is pushdown-agnostic; it just walks the tree and returns a `Some` iff it can determine which shards contain the plan's data.